### PR TITLE
[BYOM] Support for Local, Private Endpoints

### PR DESCRIPTION
**Pull request by jonathansampson**
Resolves https://github.com/brave/brave-browser/issues/39627

### Included Scenarios

_Derived from [July 8th](https://github.com/brave/brave-browser/issues/39627#issue-2396667753) and [July 18th](https://github.com/brave/brave-browser/issues/39627#issuecomment-2237540087) comments._

- [x] Warn when a private IP/endpoint is used and the feature flag is disabled.

To satisfy this requirement, a modal will be displayed upon attempting to save a model configuration with a private endpoint, while the feature is disabled. The modal will give a brief explanation of the issue, and provide guidance on how to enable the feature.

image

- [x] Warn when a private IP/endpoint is used and the feature flag is enabled (with additional context about risks).

This requirement is addressed with a 🔓 icon (and label) displayed prominently beneath the server endpoint input component. This warning element is only displayed when the server endpoint is valid only due to the optional feature having been enabled.

image

- [x] Warn the user when they attempt to use a model that has an invalid endpoint. Do not permit messages to be sent under this scenario.

This pull request introduces a new error message for the chat context. If the model's server endpoint is invalid, the user will be informed. Furthermore, a "Configure" button is displayed to assist the user in reaching the model configuration UI.

image

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

#### Feature Disabled Scenario

- Verify the feature is NOT enabled via `brave://flags/#brave-ai-chat-allow-private-ips`
- Add a local network endpoint (e.g., `http://10.198.1.12:11434/v1/chat/completions`) in a new or existing model configuration.
  - You should see a warning about the endpoint being invalid (i.e., "Non-local endpoints…")
  - You should see a modal dialog (i.e., "Private IPs Not Allowed") upon clicking "Save model"

#### Feature Enabled Scenario
- Verify the feature IS enabled
- Add a local network endpoint in a new or existing model configuration
  - You should see a warning beneath the endpoint (i.e., "🔓 This endpoint is…")

#### Feature Disabled After Having Been Enabled

- Verify that there is a properly configured model with a private endpoint
- Verify that the feature has afterwards been disabled
- Attempt to use the model in a Leo chat instance
  - You should see a warning in the Leo chat regarding the model's endpoint (i.e., "This model has an invalid endpoint…")

### CODE REVIEW

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/ai_chat/core/browser/model_validator.cc [hunk 1]
```diff
@@ -5,14 +5,73 @@
...
+
+  // Ensure the hostname is valid (non-empty and doesn't start with a dot)
+  if (host.empty() || host.front() == '.') {
+    return false;
+  }
+
+  std::string lower_host = base::ToLowerASCII(host);
+
+  // Check if the host ends with ".local"
+  if (base::EndsWith(lower_host, ".local", base::CompareCase::SENSITIVE)) {
```
Comment (thypon) [inline components/ai_chat/core/browser/model_validator.cc:32]: `.local` addresses might resolve to external hosts. If it's `.local` we should first resolve, and then replace the `.local` address to the resolved local address, at that particular point in time

**Review by thypon — CHANGES_REQUESTED** [review]
We should probably just remove the `.local` check for the moment, or improve by resolving on the UI side and passing only storing the final IP

File: components/ai_chat/core/browser/model_validator.cc (hunk 1 above)
Comment (jonathansampson) [reply components/ai_chat/core/browser/model_validator.cc:32]: Ad-hoc resolution is a bit out of my comfort zone right now, following a cursory review of `net::HostResolver`. By default, Leo supports external hosts, so should we be too concerned if a `*.local` address resolves to an external IP address? If this is an important issue, the easiest path right now might be to simply defer support for `*.local` addresses to a later date, or trust that the user has verified their DNS configuration. Worth noting that all of this is behind the experimental flag, so no out-of-the-box support for `*.local` domains is being introduced either way.

Comment (thypon) [reply components/ai_chat/core/browser/model_validator.cc:32]: I'd rather defer `*.local` resolutions to a later stage, indeed. The point is that the behavior on `.local` addresses is not stable, and depends on how it's resolved. macOS does indeed have a stable-ish resolution process; it's not the case for Microsoft or Linux-based OSes.

I'd rather dig a bit more in the Chromium source code, to see if somebody already tried to generalize this link-local resolution across OSes.

Comment (jonathansampson) [reply components/ai_chat/core/browser/model_validator.cc]: I've [removed](https://github.com/brave/brave-core/pull/26475/commits/f29cbd4ecc42f4b423e2564f3ba24cbbf753ed97) support for `.local` domains for now, and will track the effort moving forward with [#42367](https://github.com/brave/brave-browser/issues/42367).

File: components/ai_chat/core/browser/model_validator.h [hunk 2]
```diff
@@ -43,7 +43,9 @@ class ModelValidator {
   static bool IsValidContextSize(const std::optional<int32_t>& size);
   static bool HasValidContextSize(const mojom::CustomModelOptions& options);
 
-  static bool IsValidEndpoint(const GURL& endpoint);
+  static bool IsValidEndpoint(
+      const GURL& endpoint,
+      std::optional<bool> check_as_private_ip = std::nullopt);
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/model_validator.h:48]: Hey this is a mega nit, but what do you think about calling this param `allow_private_ips`?

**Review by thypon — APPROVED** [review]

**Review by fallaciousreasoning — APPROVED** [review]

File: components/ai_chat/core/browser/model_validator.h (hunk 2 above)
Comment (fallaciousreasoning) [reply components/ai_chat/core/browser/model_validator.h:48]: Happy to leave it :smile:

**Pull request merged**